app/src/engineering/kotlin/com/hedvig/app/feature/impersonation/ImpersonationReceiverActivity.kt

```diff
@@ -112,37 +109,21 @@ class ImpersonationReceiverViewModel(
   private val _state = MutableStateFlow<ViewState>(ViewState.Loading)
   val state = _state.asStateFlow()
 
-  object Event
+  object GoToLoggedInActivityEvent
 
-  private val _events = Channel<Event>(Channel.UNLIMITED)
+  private val _events = Channel<GoToLoggedInActivityEvent>(Channel.UNLIMITED)
   val events = _events.receiveAsFlow()
 
   init {
     viewModelScope.launch {
-      if (authenticationTokenService.authenticationToken == null) {
-        authenticationTokenService.authenticationToken = "123"
-      }
-      when (
-        val result = apolloClient
-          .mutation(ExchangeTokenMutation(exchangeToken))
-          .safeExecute()
-          .toEither()
-      ) {
-        is Either.Left -> {
-          _state.value = ViewState.Error(result.value.message)
-        }
-        is Either.Right -> {
-          val newToken = result.value.exchangeToken.asExchangeTokenSuccessResponse?.token
-          if (newToken == null) {
-            _state.value = ViewState.Error("Did not receive token")
-            return@launch
-          }
-          authenticationTokenService.authenticationToken = newToken
-          loginStatusService.isLoggedIn = true
+      when (val result = networkAuthRepository.exchange(AuthorizationCodeGrant(exchangeToken))) {
+        is AuthTokenResult.Error -> _state.update { ViewState.Error(result.message) }
+        is AuthTokenResult.Success -> {
+          authTokenService.updateTokens(result.accessToken, result.refreshToken)
           featureManager.invalidateExperiments()
-          _state.value = ViewState.Success
+          _state.update { ViewState.Success }
           delay(500.milliseconds)
-          _events.send(Event)
+          _events.send(GoToLoggedInActivityEvent)
         }
```

Comment:
If I understood correctly from iOS impl of impersonation, this should just be enough. Take the token from the URL as we did before, and pass it as an AuthorizationCodeGrant in the `exchange` and we should be good I think.

Reply:
Yes!